src/common/walFilter/versions/recordProcessGPDB7.c

```diff
@@ -0,0 +1,460 @@
+            {
+                THROW_FMT(FormatError,
+                          "BKPIMAGE_HAS_HOLE not set, but hole offset %" PRIu16 " length %" PRIu16, hole_offset, hole_length);
+            }
+
+            /*
+             * cross-check that bimg_len < BLCKSZ if the IS_COMPRESSED
+             * flag is set.
+             */
+            if ((bimg_info & BKPIMAGE_IS_COMPRESSED) && bimg_len == pageSize)
+            {
+                THROW_FMT(FormatError, "BKPIMAGE_IS_COMPRESSED set, but block image length %" PRIu16, bimg_len);
+            }
+        }
+        if (fork_flags & BKPBLOCK_SAME_REL)
+        {
+            if (relFileNode == NULL)
+            {
+                THROW_FMT(FormatError, "BKPBLOCK_SAME_REL set but no previous rel");
+            }
+        }
+        else
+        {
+            relFileNode = (RelFileNode *) (ptr + offset);
+            lstAdd(result, relFileNode);
+            offset += sizeof(RelFileNode);
+        }
+        offset += sizeof(BlockNumber);
+    }
+#undef COPY_HEADER_FIELD
+
+    if (mainDataSize)
+    {
+        ASSERT(record->xl_tot_len > mainDataSize);
+        // Main data is always the last one
+        const void *const mainDataPtr = ((uint8 *) record) + record->xl_tot_len - mainDataSize;
+        const RelFileNode *const mainDataRelFileNode = getRelFileNodeFromMainData(record, mainDataPtr);
+        if (mainDataRelFileNode)
+            lstAdd(result, mainDataRelFileNode);
+    }
+
+    FUNCTION_LOG_RETURN(LIST, result);
+}
+
+static void
+overrideXLogRecordBody(XLogRecordGPDB7 *const record)
+{
+    uint8 *recordData = XLogRecordData(record);
+    if (record->xl_tot_len - sizeof(XLogRecordGPDB7) - sizeof(XLogRecordDataHeaderShort) <= UINT8_MAX)
+    {
+        *((XLogRecordDataHeaderShort *) recordData) = (XLogRecordDataHeaderShort){
+            XLR_BLOCK_ID_DATA_SHORT,
+            (uint8) (record->xl_tot_len - sizeof(XLogRecordGPDB7) - sizeof(XLogRecordDataHeaderShort))
+        };
+    }
+    else
+    {
+        *((XLogRecordDataHeaderLong *) recordData) = (XLogRecordDataHeaderLong){
+            XLR_BLOCK_ID_DATA_LONG,
+            (uint32) (record->xl_tot_len - sizeof(XLogRecordGPDB7) - sizeof(XLogRecordDataHeaderLong))
+        };
+    }
+}
+
+FN_EXTERN void
+filterRecordGPDB7(XLogRecordBase *const recordBase, const PgPageSize pageSize)
+{
+    FUNCTION_LOG_BEGIN(logLevelTrace);
+        FUNCTION_LOG_PARAM(INT, (int) pageSize);
+    FUNCTION_LOG_END();
+
+    XLogRecordGPDB7 *const record = (XLogRecordGPDB7 *const) recordBase;
+
+    if (record->xl_rmid == RM_XLOG_ID && record->xl_info == XLOG_NOOP)
+        goto end;
+
+    List *const nodes = getRelFileNodes(record, pageSize);
+    if (lstEmpty(nodes)){
+        lstFree(nodes);
+        goto end;
+    }
+
+    bool hasPassFilter = false;
+    bool hasNotPassFilter = false;
+    List *notPassFilterList = lstNewP(sizeof(RelFileNode));
+    for (uint32 i = 0; i < lstSize(nodes); i++)
```

Comment:
unsigned int is always used in pgbackrest for-s with lstSize(), because lstSize() return unsigned int